DS-3140 METSRightsCrosswalk NPE During AIP Restore - No Anonymous Read

mjmarttila opened the pull request:
Consolidation/refactoring of the METSRightsCrosswalk ingest() method logic to address NPEs
and incorrect behavior when importing certain objects (e.g. a collection
with an rpName and association to a non-Admin/non-Anonymous group).

Ticket: https://jira.duraspace.org/browse/DS-3140

mjmarttila committed:
METSRightsCrosswalk ingest(...) Policy Logic Refactoring

Consolidation/refactoring of the ingest() method logic to address NPEs
and incorrect behavior when importing certain objects (e.g. a collection
with an rpName and association to a non-Admin/non-Anonymous group).

mjmarttila committed:
Typo fix

Removed extra semicolons.

terrywbrady commented:
Note: this PR was developed/tested at Georgetown and it resolved an AIP ingest issue locally.

tdonohue commented on dspace-api/src/main/java/org/dspace/content/crosswalk/METSRightsCrosswalk.java:
```diff
@@ -471,85 +471,31 @@ else if (element.getName().equals("Context"))
-                            }
-                        }
-                        policies.add(rp);
-                    } catch (NullPointerException ex) {
```
Unclear to me why this entire `for` loop has been stripped out?  Is some other area of the class achieving this same logic?

mjmarttila replied:
That `for` loop's permissions-decoding logic should be covered in the call to the `parsePermissions()` method (see the call on [line 595](https://github.com/Georgetown-University-Libraries/DSpace/blob/b387bca5e532b0d35fc6bdb31549410d6f6c2cb7/dspace-api/src/main/java/org/dspace/content/crosswalk/METSRightsCrosswalk.java#L595) and the method definition on [line 689)](https://github.com/Georgetown-University-Libraries/DSpace/blob/b387bca5e532b0d35fc6bdb31549410d6f6c2cb7/dspace-api/src/main/java/org/dspace/content/crosswalk/METSRightsCrosswalk.java#L689). I could not think of a scenario where a METS file context element would have more than one permissions child element (but folks more experienced with the internals, please speak up if I'm off-base here), and found that the looping logic would even sometimes cause an NPE when it looked at a non-permissions child element (see the [example context element](https://jira.duraspace.org/browse/DS-3140?focusedCommentId=49773&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-49773) provided in the corresponding JIRA ticket). So instead of looping through all of the child elements, I grabbed just the permissions child element (see [line 480](https://github.com/Georgetown-University-Libraries/DSpace/blob/b387bca5e532b0d35fc6bdb31549410d6f6c2cb7/dspace-api/src/main/java/org/dspace/content/crosswalk/METSRightsCrosswalk.java#L480)) and passed that to the `parsePermissions()` method.

tdonohue replied:
Hi @mjmarttila : I understand now. Thanks for the clear explanation

Technically, according to the [METSRights XSD](https://www.loc.gov/standards/rights/METSRights.xsd), a `<Context>` element can have multiple `<Permissions>` elements. (See the `<xs:complexType name="ContextTYPE">` in that XSD).

However, I have never seen an example of why this would be.  Even the examples provided for METSRights only ever have one: http://www.loc.gov/standards/rights/

So, considering this option works and is more simplistic, it seems reasonable enough. I don't know why you'd ever want more than one of these `<Permissions>` tags.

However, I think it would be worth adding comments to this effect in your changes... So, perhaps adding to line 480 a comment like `// Note: we are assuming here there will only ever be ONE <Permissions> element. Currently there a no known use cases for multiple.`

mjmarttila replied:
Hi @tdonohue,

Good catch on the METSRights specification. I think the clarification comment makes sense and I just added one in-line with your suggestion to line 480.

Thanks for the recommendation and fact-checking.

mjmarttila committed:
Comment clarity for <Permissions> element

Added comment clarifying that the code makes the assumption that a
<Context> element will only have one <Permissions> child element in a
DSpace objects METSRights file even though the specification says that
multiple permissions objects are technically possible. Thanks to
tdonohue for the recommendation.

tdonohue commented on dspace-api/src/main/java/org/dspace/content/crosswalk/METSRightsCrosswalk.java:
```diff
@@ -638,11 +586,19 @@ else if(PERSON_CONTEXTCLASS.equals(contextClass))
+                	policies.add(rp);
+                	assignPermissions(context, dso, policies);
+                } catch (NullPointerException ex) {
+                    java.util.logging.Logger.getLogger(METSRightsCrosswalk.class.getName()).log(Level.SEVERE, null, ex);
```
Another silly question: Why are we catching a NullPointerException here? Is there a check for possible null value that is missing?

It'd be good to have an explanation. If it is necessary, I'd recommend we comment this, as usually we don't see catches for NPEs (we just try to avoid null values).

It's also worth noting that you don't need the complex logger line of code. It could be simplified to just `log.error([explanation], ex);`.  There's already a global reference for a logger at [line 64](https://github.com/Georgetown-University-Libraries/DSpace/blob/05d85843e90d23ace5cad231e05ccb5fa3096eeb/dspace-api/src/main/java/org/dspace/content/crosswalk/METSRightsCrosswalk.java#L64)

mjmarttila committed:
Remove Unused NPE Try/Catch and Null Check Permissions Element

Remove legacy try/catch statement, validate that permissions element is
not null, and update log call.

mjmarttila commented:
Hi @tdonohue,

Another good catch with the NPE try/catch. I ported it over with some of the original logic but after reviewing the logic a second time, I can't see any good reason why it's necessary. I've removed the entire try/catch statement (note that I shifted the contained code one tab to the left so github is making the commit changes look more significant than they really are).

Per your second point, I also added a null check of the permissions child element (see [line 482](https://github.com/DSpace/DSpace/pull/1396/commits/20a5743b07c8c72452a3f8fddf6e6fbe88bb0afa#diff-72b9464a2e417bd31d73a338a7f3cc6eR482)) in the rare case of a bad METSRights XML file.

And finally, I updated the logger code (see [line 501](https://github.com/DSpace/DSpace/pull/1396/commits/20a5743b07c8c72452a3f8fddf6e6fbe88bb0afa#diff-72b9464a2e417bd31d73a338a7f3cc6eR501)) to use the simpler method call.

Thanks for your suggestions.

terrywbrady commented:
+1, tested locally

tdonohue commented:
+1 was able to verify the bug on current master by creating a Collection with READ access restricted to a custom group, exporting the AIP, and then attempting to restore it via AIP.   This PR fixes the NPE and the restoration process of such a collection now works. It also doesn't seem to negatively affect restores of other objects (like Items, etc).

Sorry for not getting to testing this sooner. Thanks @mjmarttila for all your hard work here!  I'll merge this shortly as it's now at +2 approval.

mjmarttila commented:
Thanks for testing and merging, @tdonohue!
